perception/autoware_tensorrt_bevdet/src/preprocess_plugin.cu

```diff
@@ -0,0 +1,347 @@
+  return 1;
+}
+
+DataType PreprocessPlugin::getOutputDataType(
+  int32_t index, DataType const * inputTypes, int32_t nbInputs) const noexcept
+{
+  // return DataType::kHALF;
+  return DataType::kFLOAT;
+}
+
+DimsExprs PreprocessPlugin::getOutputDimensions(
+  int32_t outputIndex, const DimsExprs * inputs, int32_t nbInputs,
+  IExprBuilder & exprBuilder) noexcept
+{
+  int input_h = inputs[0].d[2]->getConstantValue();
+  int input_w = inputs[0].d[3]->getConstantValue() * 4;
+
+  int output_h = input_h * m_.resize_radio - m_.crop_h;
+  int output_w = input_w * m_.resize_radio - m_.crop_w;
+
+  DimsExprs ret;
+  ret.nbDims = inputs[0].nbDims;
+  ret.d[0] = inputs[0].d[0];
+  ret.d[1] = inputs[0].d[1];
+  ret.d[2] = exprBuilder.constant(output_h);
+  ret.d[3] = exprBuilder.constant(output_w);
+
+  return ret;
+}
+
+bool PreprocessPlugin::supportsFormatCombination(
+  int32_t pos, const PluginTensorDesc * inOut, int32_t nbInputs, int32_t nbOutputs) noexcept
+{
+  bool res;
+  switch (pos) {
+    case 0:  // images
+      res = (inOut[0].type == DataType::kINT8 || inOut[0].type == DataType::kINT32) &&
+            inOut[0].format == TensorFormat::kLINEAR;
+      break;
+    case 1:  // mean
+      res = (inOut[1].type == DataType::kFLOAT) && inOut[1].format == TensorFormat::kLINEAR;
+      break;
+    case 2:  // std
+      res = (inOut[2].type == DataType::kFLOAT) && inOut[2].format == TensorFormat::kLINEAR;
+      break;
+    case 3:  // output img tensor
+      res = (inOut[3].type == DataType::kFLOAT || inOut[3].type == DataType::kHALF) &&
+            inOut[3].format == inOut[0].format;
+
+      // res = inOut[3].type == DataType::kHALF && inOut[3].format == inOut[0].format;
+      break;
+    default:
+      res = false;
+  }
+  return res;
+}
+
+void PreprocessPlugin::configurePlugin(
+  const DynamicPluginTensorDesc * in, int32_t nbInputs, const DynamicPluginTensorDesc * out,
+  int32_t nbOutputs) noexcept
+{
+  return;
+}
+
+size_t PreprocessPlugin::getWorkspaceSize(
+  const PluginTensorDesc * inputs, int32_t nbInputs, const PluginTensorDesc * outputs,
+  int32_t nbOutputs) const noexcept
+{
+  return 0;
+}
+
+int32_t PreprocessPlugin::enqueue(
+  const PluginTensorDesc * inputDesc, const PluginTensorDesc * outputDesc,
+  const void * const * inputs, void * const * outputs, void * workspace,
+  cudaStream_t stream) noexcept
+{
+  int n_img = inputDesc[0].dims.d[0];
+  int src_img_h = inputDesc[0].dims.d[2];
+  int src_img_w = inputDesc[0].dims.d[3] * 4;
+
+  int dst_img_h = outputDesc[0].dims.d[2];
+  int dst_img_w = outputDesc[0].dims.d[3];
+
+  int src_row_step = src_img_w;
+  int dst_row_step = dst_img_w;
+
+  int src_img_step = src_img_w * src_img_h * 3;
+  int dst_img_step = dst_img_w * dst_img_h * 3;
+
+  float offset_h = m_.crop_h / m_.resize_radio;
+  float offset_w = m_.crop_w / m_.resize_radio;
+
+  dim3 grid(DIVUP(dst_img_h * dst_img_w * n_img, NUM_THREADS));
+  dim3 block(NUM_THREADS);
+
+  switch (int(outputDesc[0].type)) {
```

Comment:
The `switch` statement handles `DataType::kFLOAT` and `DataType::kHALF`. Consider adding a `default` case that throws an exception or returns an error code instead of just logging an error.  This will improve robustness by handling unexpected situations more gracefully.